bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,1049 @@
+        }
+
+        public Reader contentStreamReader() {
+            return new InputStreamReader(new ByteArrayInputStream(buffer, 0, usedSize), StandardCharsets.UTF_8);
+        }
+
+        public void reset() {
+            usedSize = 0;
+        }
+    }
+
+    /**
+     * Adapter for listening to SSE event stream events.
+     *
+     * It receives the incoming text lines. Receipt of the first data line causes the CompletableFuture to complete. It
+     * then parses subsequent data according to the SSE specification. If the line starts with a 'data:' message, it
+     * adds the data to the list of strings. And if the line is empty (i.e. the last line of an event), it passes the
+     * full set of strings to the owner via a call-back method.
+     *
+     * The stream must be permanently connected, so it ignores onIdleTimeout() events.
+     *
+     * The parent class handles most fatal errors, but since the event stream is supposed to be permanently connected,
+     * the following events are also considered as fatal:
+     *
+     * <li>onClosed()</li>
+     * <li>onReset()</li>
+     */
+    private class EventStreamListenerAdapter extends BaseStreamListenerAdapter<Boolean> {
+        private final DataFrameCollector eventData = new DataFrameCollector();
+
+        @Override
+        public void onClosed(@Nullable Stream stream) {
+            handleHttp2Error(Http2Error.CLOSED);
+        }
+
+        @Override
+        public void onData(@Nullable Stream stream, @Nullable DataFrame frame, @Nullable Callback callback) {
+            Objects.requireNonNull(frame);
+            Objects.requireNonNull(callback);
+            synchronized (this) {
+                eventData.append(frame.getData());
+                BufferedReader reader = new BufferedReader(eventData.contentStreamReader());
+                @SuppressWarnings("null")
+                List<String> receivedLines = reader.lines().collect(Collectors.toList());
+
+                // a blank line marks the end of an SSE message
+                boolean endOfMessage = !receivedLines.isEmpty()
+                        && receivedLines.get(receivedLines.size() - 1).isBlank();
+
+                if (endOfMessage) {
+                    eventData.reset();
+                    // receipt of ANY message means the event stream is established
+                    if (!completable.isDone()) {
+                        completable.complete(Boolean.TRUE);
+                    }
+                    // append any 'data' field values to the event message
+                    StringBuilder eventContent = new StringBuilder();
+                    for (String receivedLine : receivedLines) {
+                        if (receivedLine.startsWith("data:")) {
+                            eventContent.append(receivedLine.substring(5).stripLeading());
+                        }
+                    }
+                    // note: StringBuilder.isEmpty() only available from Java 15+
```

Comment:
I am running this PR 7/24 on my operative OH v3.4.x system. 'Eat your own dog food'. So every change that I make is done first on private my 3.4.x branch and tested thereon. And if it passes the v3.4.x test, then I push the exact same code to the OH v4.0 branch of this PR. This means that the code base MUST be Java 11 compatible. Otherwise it will be unproven code not actually tested on a live operative system. Which I think could be rather risky.

Please keep in mind that I use Maven to build the test Jars. So even if the v 4.0 code is 'only' Java 11 compatible, the resulting OH v4.0 Jar is built on Java 17, whereas the respective OH v3.4.x Jar is built on Java 11.

One day, I shall probably migrate my operative live system to OH v4.0. But I will probably wait one or two more monthly milestones before I do so. But once I have done so, then I plan to drop the v3.4.2 branch.

Also please note that there is NOTHING WRONG with code that does (say) an 'instanceof' check and type cast on two lines. This is still absolutely and 100% perfectly valid Java 17 syntax. Yeah sure the additional one line check and cast is neater. But the prior syntax will NEVER be dropped.

Probably one day once I make the operative switch to OH v4.0 I will go back through the old code and adopt the newer syntax. But not now. .. yet if it makes you feel happier, I could now pre-annotate (with 'TODO') every line where such an adoption could eventually be foreseen. (i.e. make it visible) => WDYT?

Reply:
That would generate checkstyle warnings (TodoCommentCheck). Since you are not planning to use Java 17 features anytime soon, this would potentially block a merge before M3 or M4 to avoid introducing new checkstyle warnings. As you said, Java 11 is perfectly fine, so there's no need to accumulate such TODO's/FIXME's. After the merge, anyone can of course create new PRs that would use such features.

For the particular code comment above, please remove it as it's not relevant, or leave this PR comment open until removed to not forget.